### PR TITLE
centralize plugin options

### LINKED ISSUE #6376: Centralize each plugin's option processing to the top of the function.
As of https://github.com/babel/babel/pull/6350, plugins have changes from

```
module.exports = function(babel){
```
to
```
module.exports = function(babel, options){
```

Now that the options are easily accessible in the top-level scope, alongside the old `this.opts` and `state.opts` methods, it would help readability and validation if we move the accesses to these options to the top of the plugin functions.

In some places the references are already nicely grouped, like https://github.com/babel/babel/blob/91cde0148d16afe9f9192953084f0b668885cdf4/packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js#L93

so they are easy to move.

Some places they are generally scattered around the plugins, like in https://github.com/babel/babel/blob/91cde0148d16afe9f9192953084f0b668885cdf4/packages/babel-plugin-transform-runtime/src/index.js#L111

Once everything is centralized at the top, we can start adding validation for the config values to give better feedback to users if they pass unknown options or invalid values for the options. That said, it is up to whoever takes this on for whether they want to tackle the validation too.

### PR DESCRIPTION
**Pull request by RusinovAnton**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues             | Fixes #6376 
| Patch: Bug Fix?          | No (?)
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added/Pass?        | No
| Spec Compliancy?         | 
| License                  | MIT
| Doc PR                   |
| Any Dependency Changes?  | No

Refactor plugins code to centralize options usage as requested in #6376

### CODE REVIEW

**Comment by RusinovAnton** [comment]
@loganfsmyth please review

File: packages/babel-plugin-transform-react-constant-elements/src/index.js [hunk 1]
```diff
@@ -71,8 +72,8 @@ export default function({ types: t }) {
         // This transform takes the option `allowMutablePropsOnTags`, which is an array
         // of JSX tags to allow mutable props (such as objects, functions) on. Use sparingly
         // and only on tags you know will never modify their own props.
-        if (this.opts.allowMutablePropsOnTags != null) {
-          if (!Array.isArray(this.opts.allowMutablePropsOnTags)) {
+        if (allowMutablePropsOnTags != null) {
+          if (!Array.isArray(allowMutablePropsOnTags)) {
             throw new Error(
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-react-constant-elements/src/index.js]: Looks like this has some validation that should be moved too.

File: packages/babel-plugin-transform-runtime/src/index.js [hunk 2]
```diff
@@ -15,13 +17,11 @@ export default function({ types: t }) {
...
-      const moduleName = getRuntimeModuleName(this.opts);
+      const moduleName = getRuntimeModuleName(options);
 
-      if (this.opts.helpers !== false) {
-        const baseHelpersDir = this.opts.useBuiltIns
-          ? "helpers/builtin"
-          : "helpers";
-        const helpersDir = this.opts.useESModules
+      if (helpers !== false) {
+        const baseHelpersDir = useBuiltIns ? "helpers/builtin" : "helpers";
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-runtime/src/index.js]: This and the next few lines are all validation for the options, so that should get moved up too.

File: packages/babel-plugin-transform-runtime/src/index.js [hunk 3]
```diff
@@ -41,7 +41,7 @@ export default function({ types: t }) {
         });
       }
 
-      if (this.opts.polyfill && this.opts.useBuiltIns) {
+      if (polyfill && useBuiltIns) {
         throw new Error(
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-runtime/src/index.js]: This validation can be moved with the options.

File: packages/babel-plugin-transform-runtime/src/index.js [hunk 4]
```diff
@@ -15,13 +17,11 @@ export default function({ types: t }) {
 
   return {
     pre(file) {
-      const moduleName = getRuntimeModuleName(this.opts);
+      const moduleName = getRuntimeModuleName(options);
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-runtime/src/index.js]: This function has `options` in scope, so it doesn't need the argument at all. Can also move the `opts.moduleName` reference into your destructuring.

File: packages/babel-plugin-transform-react-jsx/src/index.js [hunk 5]
```diff
@@ -22,7 +23,7 @@ export default function({ types: t }) {
 
   visitor.Program = function(path, state) {
     const { file } = state;
-    let id = state.opts.pragma || "React.createElement";
+    let id = pragma || "React.createElement";
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-react-jsx/src/index.js]: This default assignment can get moved to the top too.

File: packages/babel-plugin-transform-optional-chaining/src/index.js [hunk 6]
```diff
@@ -123,7 +125,7 @@ export default function({ types: t }) {
           return;
         }
 
-        optional(path, findReplacementPath(path), this.opts.loose);
+        optional(path, findReplacementPath(path), loose);
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-optional-chaining/src/index.js]: `options` is in scope of the `optional` function, so this third argument isn't needed anymore.

File: packages/babel-plugin-transform-es2015-template-literals/src/index.js [hunk 7]
```diff
@@ -60,7 +62,7 @@ export default function({ types: t }) {
         }
 
         let helperName = "taggedTemplateLiteral";
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-template-literals/src/index.js]: This helper name construction can happen at the top now since it's constant.

File: packages/babel-plugin-transform-es2015-modules-systemjs/src/index.js [hunk 8]
```diff
@@ -359,7 +360,7 @@ export default function({ types: t }) {
           path.node.body = [
             buildTemplate({
               SYSTEM_REGISTER: t.memberExpression(
-                t.identifier(state.opts.systemGlobal || "System"),
+                t.identifier(systemGlobal || "System"),
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-modules-systemjs/src/index.js]: This `"System"` default might as well be moved to the top too.

File: packages/babel-plugin-transform-es2015-classes/src/index.js [hunk 9]
```diff
@@ -51,7 +53,7 @@ export default function({ types: t }) {
         node[VISITED] = true;
 
         let Constructor = VanillaTransformer;
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-classes/src/index.js]: This `Constructor` assignment is can be moved to the top since it just uses `loose`.

File: packages/babel-plugin-transform-class-properties/src/index.js [hunk 10]
```diff
@@ -52,8 +53,8 @@ export default function({ types: t }) {
     inherits: syntaxClassProperties,
 
     visitor: {
-      Class(path, state) {
-        const buildClassProperty = state.opts.loose
+      Class(path) {
+        const buildClassProperty = loose
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-class-properties/src/index.js]: This whole assignment can get moved to the top since it only cares about `loose`.

**Comment by loganfsmyth** [comment]
The test failures are related to your change, but aren't because of anything you broke. As long as they pass locally for you, don't worry too much.

I'll look into seeing what we can be about the failures.

**Comment by RusinovAnton** [comment]
@loganfsmyth thanks a lot for your response. I'll work on changes tomorrow

**Comment by xtuc** [comment]
Thanks @RusinovAnton, that's an amazing first time contribution!

File: packages/babel-plugin-transform-runtime/src/index.js [hunk 11]
```diff
@@ -2,10 +2,14 @@ import { addDefault, isModule } from "babel-helper-module-imports";
 
 import definitions from "./definitions";
 
-export default function({ types: t }) {
-  function getRuntimeModuleName(opts) {
```
Comment (RusinovAnton) [inline packages/babel-plugin-transform-runtime/src/index.js]: I've also removed `getRuntimeModuleName` function since is redundant now.

**Comment by RusinovAnton** [comment]
Actually, I've unintentionally lied about tests. They're not passed atm.
```
make test
./node_modules/.bin/eslint scripts packages codemods *.js --format=codeframe --rulesdir="./eslint_rules"
./scripts/test.sh
/Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:80
throw new Error("Module build failed: TypeError: [BABEL] /Users/rusinov/my-projects/babel/packages/babel-standalone/src/index.js: Cannot read property 'loose' of undefined (While processing: \"base$2\")\n    at _default (/Users/rusinov/my-projects/babel/packages/babel-plugin-transform-optional-chaining/lib/index.js:12:31)\n    at /Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:239:12\n    at cachedFunction (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/caching.js:48:17)\n    at loadPluginDescriptor (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:269:28)\n    at /Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:56:14\n    at Array.map (<anonymous>)\n    at OptionManager.mergeOptions (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:55:34)\n    at /Users/rusinov/my-projects/babel/packages/b

Error: Module build failed: TypeError: [BABEL] /Users/rusinov/my-projects/babel/packages/babel-standalone/src/index.js: Cannot read property 'loose' of undefined (While processing: "base$2")
    at _default (/Users/rusinov/my-projects/babel/packages/babel-plugin-transform-optional-chaining/lib/index.js:12:31)
    at /Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:239:12
    at cachedFunction (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/caching.js:48:17)
    at loadPluginDescriptor (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:269:28)
    at /Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:56:14
    at Array.map (<anonymous>)
    at OptionManager.mergeOptions (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:55:34)
    at /Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:83:15
    at Array.forEach (<anonymous>)
    at OptionManager.mergeOptions (/Users/rusinov/my-projects/babel/packages/babel-core/lib/config/option-manager.js:82:15)
    at Object.<anonymous> (/Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:80:7)
    at __webpack_require__ (/Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:30:30)
    at /Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:73:18
    at /Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:76:10
    at webpackUniversalModuleDefinition (/Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:3:20)
    at Object.<anonymous> (/Users/rusinov/my-projects/babel/packages/babel-standalone/babel.js:10:3)
    at Module._compile (module.js:624:30)
    at Module._compile (/Users/rusinov/my-projects/babel/node_modules/pirates/lib/index.js:88:24)
    at Module._extensions..js (module.js:635:10)
    at Object.newLoader [as .js] (/Users/rusinov/my-projects/babel/node_modules/pirates/lib/index.js:93:7)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/rusinov/my-projects/babel/packages/babel-standalone/test/babel.js:2:15)
    at Module._compile (module.js:624:30)
    at Module._compile (/Users/rusinov/my-projects/babel/node_modules/pirates/lib/index.js:88:24)
    at Module._extensions..js (module.js:635:10)
    at Object.newLoader [as .js] (/Users/rusinov/my-projects/babel/node_modules/pirates/lib/index.js:93:7)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at /Users/rusinov/my-projects/babel/node_modules/mocha/lib/mocha.js:230:27
    at Array.forEach (<anonymous>)
    at Mocha.loadFiles (/Users/rusinov/my-projects/babel/node_modules/mocha/lib/mocha.js:227:14)
    at Mocha.run (/Users/rusinov/my-projects/babel/node_modules/mocha/lib/mocha.js:513:10)
    at Object.<anonymous> (/Users/rusinov/my-projects/babel/node_modules/mocha/bin/_mocha:480:18)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
    at Function.Module.runMain (module.js:665:10)
    at startup (bootstrap_node.js:201:16)
    at bootstrap_node.js:626:3
make: *** [test-only] Error 1
```

**Comment by RusinovAnton** [comment]
>  They're not passed atm.
Oh, or I guess they're not even started there?

**Comment by hzoo** [comment]
Yeah it's because babel-standalone is expecting a certain options object for plugin/presets so need to modify that logic https://github.com/babel/babel/blob/master/packages/babel-standalone/src/index.js#L43 - someone else can look into it if you can't/don't want to.

**Comment by RusinovAnton** [comment]
@hzoo, I guess thats an easy fix but isn't that out of scope of current PR?

**Comment by hzoo** [comment]
This PR is the change that causes this error so it still needs to be fixed to merge, I was just saying you don't have to do it but you'll need to wait for someone to fix it

**Comment by RusinovAnton** [comment]
@hzoo, ok, should I open issue for that?

File: packages/babel-plugin-transform-es2015-arrow-functions/src/index.js [hunk 12]
```diff
@@ -1,11 +1,12 @@
 import type NodePath from "babel-traverse";
 
-export default function() {
+export default function(babel, options) {
+  const { spec } = options;
+  const isSpec = !!spec;
```
Comment (existentialism) [inline packages/babel-plugin-transform-es2015-arrow-functions/src/index.js]: Is this change needed? Not sure it's really an improvement?

File: packages/babel-plugin-transform-es2015-classes/src/index.js [hunk 13]
```diff
@@ -3,10 +3,15 @@ import VanillaTransformer from "./vanilla";
...
 import nameFunction from "babel-helper-function-name";
 
-export default function({ types: t }) {
+export default function({ types: t }, options) {
+  const { loose } = options;
+
   // todo: investigate traversal requeueing
   const VISITED = Symbol();
 
+  let Constructor = VanillaTransformer;
```
Comment (existentialism) [inline packages/babel-plugin-transform-es2015-classes/src/index.js]: Nit:

```js
const Constructor = loose ? LooseTransformer : VanillaTransformer;
```

Comment (Andarist) [inline packages/babel-plugin-transform-es2015-classes/src/index.js]: atm there are 2 code styles regarding this:
```js
// 1
const buildClassProperty = loose
  ? buildClassPropertyLoose
  : buildClassPropertySpec;
```
and
```js
let Constructor = VanillaTransformer;
if (loose) Constructor = LooseTransformer;
```

I would choose one 1 of them (my choice would be the former one) and unify it across the plugins

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 14]
```diff
@@ -8,7 +8,27 @@ import {
...
+    loose,
+    noInterop,
+    strict,
+    strictMode,
+
+    // Defaulting to 'true' for now. May change before 7.x major.
+    allowCommonJSExports = true,
+  } = options;
+
+  const ref = {
```
Comment (existentialism) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: I personally prefer if this was moved back. I think there's a difference between hoisting option parsing and usage.

File: packages/babel-plugin-transform-es2015-modules-systemjs/src/index.js [hunk 15]
```diff
@@ -22,8 +22,10 @@ const buildExportAll = template(`
 
 const TYPE_IMPORT = "Import";
 
-export default function({ types: t }) {
+export default function({ types: t }, options) {
+  const { systemGlobal } = options;
   const IGNORE_REASSIGNMENT_SYMBOL = Symbol();
+  const system = systemGlobal || "System";
```
Comment (existentialism) [inline packages/babel-plugin-transform-es2015-modules-systemjs/src/index.js]: Can set this default in the destructure above.

File: packages/babel-plugin-transform-es2015-spread/src/index.js [hunk 16]
```diff
@@ -1,9 +1,8 @@
-export default function({ types: t }) {
-  function getSpreadLiteral(spread, scope, state) {
-    if (
-      state.opts.loose &&
-      !t.isIdentifier(spread.argument, { name: "arguments" })
-    ) {
+export default function({ types: t }, options) {
+  const { loose } = options;
+
+  function getSpreadLiteral(spread, scope) {
```
Comment (existentialism) [inline packages/babel-plugin-transform-es2015-spread/src/index.js]: Can remove passing down `state` at the call sites:

https://github.com/babel/babel/pull/6381/files#diff-656ce81f115b2958c89cf5869995f057R34

File: packages/babel-plugin-transform-es2015-classes/src/index.js (hunk 13 above)
Comment (RusinovAnton) [reply packages/babel-plugin-transform-es2015-classes/src/index.js]: @Andarist 
I'm totally new to project. Could you please clarify what exactly is the former one?

File: packages/babel-plugin-transform-react-constant-elements/src/index.js [hunk 17]
```diff
@@ -1,4 +1,9 @@
-export default function transformReactConstantElement({ types: t }) {
+export default function transformReactConstantElement({ types: t }, options) {
+  const { allowMutablePropsOnTags } = options;
+  const isAllowMutablePropsOnTagsAnArray = Array.isArray(
```
Comment (existentialism) [inline packages/babel-plugin-transform-react-constant-elements/src/index.js]: This and `isAllowMutablePropsOnTagsNotNull ` are also cases where I don't think a change is needed.

File: packages/babel-plugin-transform-es2015-computed-properties/src/index.js [hunk 18]
```diff
@@ -1,4 +1,9 @@
-export default function({ types: t, template }) {
+export default function({ types: t, template }, options) {
+  const { loose: isLoose } = options;
+
+  let callback = spec;
```
Comment (Andarist) [inline packages/babel-plugin-transform-es2015-computed-properties/src/index.js]: personally I would rename `callback` to smth like `pushComputedProps`, and accordingly:
- `spec` -> `pushComputedPropsSpec`
- `loose` -> `pushComputedPropsLoose`

to give them all more meaning

File: packages/babel-plugin-transform-es2015-classes/src/index.js (hunk 13 above)
Comment (existentialism) [reply packages/babel-plugin-transform-es2015-classes/src/index.js]: @Andarist is just suggesting being consistent in the areas where we do something based on loose mode (often a template or function). FWIW, I too prefer the ternary.

File: packages/babel-plugin-transform-es2015-modules-amd/src/index.js [hunk 19]
```diff
@@ -14,34 +14,22 @@ const buildWrapper = template(`
   })
 `);
 
-export default function({ types: t }) {
+export default function({ types: t }, options) {
+  const { allowTopLevelThis, loose, noInterop, strict, strictMode } = options;
+  const ref = { allowTopLevelThis, loose, noInterop, strict, strictMode };
```
Comment (Andarist) [inline packages/babel-plugin-transform-es2015-modules-amd/src/index.js]: `ref` could be renamed to smth like `optionsCopy`, is there even need to copy it? is there danger that it gets mutated?

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 14 above)
Comment (Andarist) [reply packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: + we do not end up with `ref` variable which doesn't hold any meaningful info

File: packages/babel-plugin-transform-es2015-modules-amd/src/index.js (hunk 19 above)
Comment (Andarist) [reply packages/babel-plugin-transform-es2015-modules-amd/src/index.js]: Actually I thin the same as @existentialism here - https://github.com/babel/babel/pull/6381#discussion_r142783867

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 14 above)
Comment (RusinovAnton) [reply packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Alright, I've got your point

**Comment by hzoo** [comment]
Ok I have a fix I think @loganfsmyth, opts needs to default to `{}` since we are going to destructure in some cases (like in this PR)

File: packages/babel-plugin-transform-react-constant-elements/src/index.js [hunk 20]
```diff
@@ -71,8 +75,8 @@ export default function transformReactConstantElement({ types: t }) {
         // This transform takes the option `allowMutablePropsOnTags`, which is an array
         // of JSX tags to allow mutable props (such as objects, functions) on. Use sparingly
         // and only on tags you know will never modify their own props.
-        if (this.opts.allowMutablePropsOnTags != null) {
-          if (!Array.isArray(this.opts.allowMutablePropsOnTags)) {
+        if (allowMutablePropsOnTags != null) {
+          if (isAllowMutablePropsOnTagsAnArray) {
             throw new Error(
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-react-constant-elements/src/index.js]: This `if` and `throw` can be moved I believe.

**Comment by loganfsmyth** [comment]
Okay, I merged https://github.com/babel/babel/pull/6435, let's merge that in and see how the tests look.

**Comment by RusinovAnton** [comment]
@loganfsmyth anything I missed here?

**Comment by loganfsmyth** [comment]
Always fun when fixing one bug uncovers another. Thanks for testing our tests for us :P Apparently a few of our tests were still running against an older version of babel-core, rather than the once included in the repository, so the `options` param was `undefined` in those cases.

PR for more fixes: https://github.com/babel/babel/pull/6454

**Comment by loganfsmyth** [comment]
I merged master into this branch with the other fixes I mentioned, let's see how things look.

**Comment by loganfsmyth** [comment]
Thanks!
